### ISSUE TITLE
Add flat frame correction

### ISSUE THREAD

**Issue by ilia3101** [ilia3101/MLV-App#154]
I want to do this, I will implement the processing part. I am just so tired of dust (!!!😡😡😡)

I am not sure how it will work in GUI tho, probably same as dark frame hopefully can get some help.

Or am i dumb and it's already in there?

**Comment by masc4ii** [comment]
What is a flat frame correction? For correcting a vignette? Then we have something like that... But it has to be applied manually.

**Comment by ilia3101** [comment]
Flat field corrects any unevenness of exposure across the frame, which includes vignetting and dust particles (unless they're just completely black).

I imagine it works like this:
Creating the flat frame: average a few frames of the sky or something completely flat + adjust the exposure so average pixel value is 1 + make it black and white
To apply: Divide pixels in the normal frame by flat frame's pixels in the same location.

**Comment by ilia3101** [comment]
I mainly want to use it for these dust blobs I have

**Comment by bouncyball-git** [comment]
Also flat field processing is implemented in mlv_dump by g3gg0, so you can take a look there. I think @dannephoto has good experience of using FF.

**Comment by dannephoto** [comment]
Oh, I've been through flat framing and my experience is that it's hard to use with ease.  Not at all like darkframing. 
If you want to get rid of dust you might want to check the remove grain filter. It's very effective and works like an average median filter:
https://www.magiclantern.fm/forum/index.php?topic=15774.msg208294#msg208294

Maybe something for Mlv App?

**Comment by masc4ii** [comment]
Hm... but an 2D average median filter is already implemented in MLVApp. But wait some hours, I think I'll upload something very cool this evening! Need some testers then! 😄

**Comment by ilia3101** [comment]
> Oh, I've been through flat framing and my experience is that it's hard to use with ease. Not at all like darkframing.
> If you want to get rid of dust you might want to check the remove grain filter. It's very effective and works like an average median filter:
> https://www.magiclantern.fm/forum/index.php?topic=15774.msg208294#msg208294
> 
> Maybe something for Mlv App?

Looks like something for small scale issues, not dust blobs at f4, would maybe help for f22 but I don't ever shoot like that. However I like a1ex's idea there. I wonder what effect it will have on moire.

I have some ideas for flat framing to fix issues I can imagine. What were your main difficulties?

**Comment by masc4ii** [comment]
The median filter can remove little moiree (some single color pixels at contrast lines)... but removes some detail too (like nearly all filters). 
For some moiree, the new recursive bilateral filter works great, have a look:
_Before:_
![Bildschirmfoto 2019-03-16 um 19 37 12](https://user-images.githubusercontent.com/30245296/54479952-1aab4d00-4823-11e9-9d6a-d57868e6c58b.png)
_After:_
![Bildschirmfoto 2019-03-16 um 19 37 42](https://user-images.githubusercontent.com/30245296/54479954-239c1e80-4823-11e9-950f-3255da9384a2.png)

**Comment by dannephoto** [comment]
Wow, bilateral ftw.
Bouncyball. Firstly a custom flat frame for each lems is needed. Also needs testing regarding what f-stop to use. Not always best with same as the recorded file. Automation parameters will be difficult.

**Comment by bouncyball-git** [comment]
Yeah... I remember that lot of hoop jumping is needed to properly use flat frame and have good results.

**Comment by ilia3101** [comment]
Ok I am reconsidering a bit... also where is dark frame implemented, in llrawproc?

**Comment by masc4ii** [comment]
Yes. 
MLVApp/src/mlv/llrawproc/darkframe.c/.h
